File: nodejs/backend/src/controller/leadRequestController.ts
import { Request, Response } from 'express';
import { LeadRequestService } from '../service/LeadRequestService';
import { toLeadRequestResponseDto } from '../dto/leadRequestDto';
import { requireStringParam } from '../error/paramHelpers';

export class LeadRequestController {
  constructor(private readonly leadRequestService: LeadRequestService) {}

  create = async (req: Request, res: Response): Promise<void> => {
    const request = await this.leadRequestService.create(req.body);
    res.status(201).json(toLeadRequestResponseDto(request));
  };

  list = async (_req: Request, res: Response): Promise<void> => {
    const requests = await this.leadRequestService.list();
    res.json(requests.map(toLeadRequestResponseDto));
  };

  remove = async (req: Request, res: Response): Promise<void> => {
    await this.leadRequestService.delete(requireStringParam(req, 'id'));
    res.status(204).send();
  };
}

File: rust/backend/src/controller/lead_request_controller.rs
use crate::{
    dto::{CreateLeadRequest, LeadRequestResponse},
    error::AppError,
    state::AppState,
};
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use uuid::Uuid;

pub async fn create_request(
    State(state): State<AppState>,
    Json(payload): Json<CreateLeadRequest>,
) -> Result<impl IntoResponse, AppError> {
    let request = state.lead_request_service.create(payload).await?;
    Ok((StatusCode::CREATED, Json(LeadRequestResponse::from(request))))
}

pub async fn list_requests(
    State(state): State<AppState>,
) -> Result<impl IntoResponse, AppError> {
    let requests: Vec<LeadRequestResponse> = state
        .lead_request_service
        .list()
        .await?
        .into_iter()
        .map(LeadRequestResponse::from)
        .collect();

    Ok(Json(requests))
}

pub async fn delete_request(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<impl IntoResponse, AppError> {
    state.lead_request_service.delete(id).await?;
    Ok(StatusCode::NO_CONTENT)
}
